tests/Unit/ViewModels/FormBuilderViewModelTest.php

```diff
@@ -29,31 +29,37 @@ public function testShouldReturnStorageData()
         $mockForm = DonationForm::factory()->create();
         $formId = $mockForm->id;
 
-        $this->assertSame(
-            [
-                'resourceURL' => rest_url(FormBuilderRestRouteConfig::NAMESPACE . '/form/' . $formId),
-                'previewURL' => (new GenerateDonationFormPreviewRouteUrl())($formId),
-                'nonce' => wp_create_nonce('wp_rest'),
-                'blockData' => $mockForm->blocks->toJson(),
-                'settings' => $mockForm->settings->toJson(),
-                'currency' => give_get_currency(),
-                'formDesigns' => array_map(static function ($designClass) {
-                    /** @var FormDesign $design */
-                    $design = give($designClass);
-
-                    return [
-                        'id' => $design::id(),
-                        'name' => $design::name(),
-                    ];
-                }, give(FormDesignRegistrar::class)->getDesigns()),
-                'formPage' => [
-                    'isEnabled' => give_is_setting_enabled(give_get_option('forms_singular')),
-                    // Note: Boolean values must be nested in an array to maintain boolean type, see \WP_Scripts::localize().
-                    'permalink' => add_query_arg(['p' => $formId], site_url('?post_type=give_forms')),
-                    'rewriteSlug' => get_post_type_object('give_forms')->rewrite['slug'],
-                ],
-            ],
-            $viewModel->storageData($formId)
-        );
+        $storageData = $viewModel->storageData($formId);
+
+        foreach([
+            'formId',
+            'resourceURL',
+            'previewURL',
+            'nonce',
+            'blockData',
+            'settings',
+            'currency',
+            'formDesigns',
+            'formPage',
+            'gateways',
+            'gatewaySettingsUrl',
+            'isRecurringEnabled',
+            'recurringAddonData',
+            'emailTemplateTags',
+            'emailNotifications',
+            'emailPreviewURL',
+        ] as $key) {
+            $this->assertArrayHasKey($key, $storageData);
```

Comment:
This is a pretty big difference in testing as before we were making sure the entire array keys and values are expected - where as now we're only checking the array key.  Can you change this back to `assertSame`?

Reply:
Asserting the values in addition to the shape feels heavy handed and tightly couples the test to the implementation. Any change to the values or any addition of filters would break a fragile test like this.

My understanding is that we should be testing the _shape_ of the data, which asserting the keys accomplishes.

Reply:
When it comes to writing tests for view models, I prefer to be heavy handed so I can be confident the data given to the view is exactly what it needs to function.  In fact, that's one of the main reasons I create view models - to extract predictable data I can write tests for.  Our front-end is fragile in nature, relying on a very specific data shape with specific data values and if either one of those is off, the view won't work properly and could break the page. 

So I see changes to the values affecting the tests a huge benefit and in my experience has prevented the front-end from breaking.  If somewhere up the chain there is a discrepancy between one of those values - and the front-end is not expecting it - i'm hopeful the test would catch it first.

Reply:
Ok, so you are wanting the redundancy here as supporting a critical system. I'll revert my changes to the test.

Reply:
Resolved, see https://github.com/impress-org/givewp-next-gen/pull/181/commits/ed71b41a7652ee079df6cfbd7ff58d366b6aa7f1

Reply:
Yup, well said. Much appreciated